include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -472,46 +472,62 @@ template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
 template <typename _T>
-struct __usm_host_or_buffer_accessor
+struct __usm_or_buffer_accessor
 {
   private:
     using __accessor_t = sycl::accessor<_T, 1, sycl::access::mode::read_write, __dpl_sycl::__target_device,
                                         sycl::access::placeholder::false_t>;
     __accessor_t __acc;
     _T* __ptr = nullptr;
     bool __usm = false;
+    size_t __offset;
 
   public:
     // Buffer accessor
-    __usm_host_or_buffer_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>* __sycl_buf)
-        : __acc(sycl::accessor(*__sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{})), __usm(false)
+    __usm_or_buffer_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>* __sycl_buf)
+        : __acc(sycl::accessor(*__sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{})), __usm(false),
+          __offset(0)
+    {
+    }
+    __usm_or_buffer_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>* __sycl_buf, size_t __acc_offset)
+        : __acc(sycl::accessor(*__sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{})), __usm(false),
+          __offset(__acc_offset)
     {
     }
 
     // USM pointer
-    __usm_host_or_buffer_accessor(sycl::handler& __cgh, _T* __usm_buf) : __ptr(__usm_buf), __usm(true) {}
+    __usm_or_buffer_accessor(sycl::handler& __cgh, _T* __usm_buf) : __ptr(__usm_buf), __usm(true), __offset(0) {}
+    __usm_or_buffer_accessor(sycl::handler& __cgh, _T* __usm_buf, size_t __ptr_offset)
+        : __ptr(__usm_buf), __usm(true), __offset(__ptr_offset)
+    {
+    }
 
     auto
     __get_pointer() const // should be cached within a kernel
     {
-        return __usm ? __ptr : &__acc[0];
+        return __usm ? __ptr + __offset : &__acc[__offset];
     }
 };
 
 template <typename _ExecutionPolicy, typename _T>
-struct __usm_host_or_buffer_storage
+struct __result_and_scratch_storage
 {
   private:
     using __sycl_buffer_t = sycl::buffer<_T, 1>;
+
+    _ExecutionPolicy __exec;
+    ::std::shared_ptr<_T> __scratch_buf;
```

Comment:
Are we going really to share this fields? Or simple manage memory by this way?
I think - the second case.
But this mean `std::unique_ptr' should be enough for us.

Reply:
I think this should stay as a `shared_ptr`. There is a situation where a variable of this class is copied and therefore two objects have ownership of the pointer. This happens when a `__future` is constructed to return the result value.

So due to the possibility of non-reference passing and copying of the class I think this should stay as is.

Error when using `unique_ptr`.
```
/home/aidanbelton/oneDPL/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:147:41: error: call to implicitly-deleted copy constructor of 'oneapi::dpl::__par_backend_hetero::__result_and_scratch_storage<oneapi::dpl::execution::device_policy<> &, TestUtils::Sum>'
  147 |         return __future(__reduce_event, __scratch_container);
```

Reply:
The fix is very simple:
`return __future(__reduce_event, std::move(__scratch_container));

Reply:
Wouldn't `unique_ptr` limit what the user can do with the future-like object we return? I think `shared_ptr` is the more general approach here.

Reply:
@julianmi exactly, you are right.
But let's take a look at https://en.cppreference.com/w/cpp/thread/future/operator%3D
We haven't requirement that should be able to have copyable result.

Reply:
Thanks, @SergeyKopienko. I think we can switch to `unique_ptr` then. We should do so throughout the unified USM or buffer storage though.

Reply:
See my update: https://github.com/oneapi-src/oneDPL/pull/1354#discussion_r1562531934

Reply:
I think there is a discussion to be had regarding the exact usage of this class. However, I believe it is well beyond the scope of this PR, which is fundamentally a performance optimization. The usage of the shared_ptr existed prior to this change. I propose that this sort of architectural discussion and change should be placed into an issue or some other method of communication as it is not relevant to the proposed changes.

Reply:
Additional moment - we shouldn't broke existing behavior of this class : our users already have the code where it's copyable and moveable.